### PR TITLE
Proposal admin actions

### PR DESCRIPTION
**Pull request by zemogle**
Adding some admin actions to:

- Make selected proposals public
- Rollover timeallocations to the next semester (but only if there is an allocation in the current semester and there isn't already an allocation in the next semester)

I've added test coverage.

### CODE REVIEW

File: observation_portal/proposals/admin.py [hunk 1]
```diff
@@ -111,6 +114,46 @@ def get_actions(self, request):
...
+
+    @admin.action(description='Rollover time allocation for selected proposals')
+    def rollover_selected(self, request, queryset):
+        now = timezone.now()
+        currentsemester = Semester.objects.filter(start__lte=now, end__gte=now).first()
+        nextsemester = Semester.objects.filter(start__gte=currentsemester.end).order_by('start').first()
+        if 'apply' in request.POST:
+            for obj in queryset:
+                allocations = obj.timeallocation_set.filter(semester=currentsemester)
+                if obj.timeallocation_set.filter(semester=nextsemester):
```
Comment (jnation3406) [inline observation_portal/proposals/admin.py:130]: you should explicitly call .exists() on the queryset here

File: observation_portal/proposals/admin.py [hunk 2]
```diff
@@ -111,6 +114,46 @@ def get_actions(self, request):
...
+                    continue
+                for allocation in allocations:
+                    newtime = TimeAllocation.objects.create(semester=nextsemester, proposal=obj, instrument_types=allocation.instrument_types)
+                    newtime.std_allocation = allocation.std_allocation - allocation.std_time_used
```
Comment (jnation3406) [inline observation_portal/proposals/admin.py:134]: Do we need any checks here to ensure we aren't setting negative allocations? Like a max(allocation.std_allocation - allocation.std_time_used, 0) perhaps?

File: observation_portal/proposals/admin.py [hunk 3]
```diff
@@ -111,6 +114,46 @@ def get_actions(self, request):
...
+                    newtime.save()
+
+            self.message_user(request, 'Successfully rolled over time for {} proposal(s)'.format(queryset.count()))
+            return HttpResponseRedirect(request.get_full_path())
+        
+        proposals = []
+        rejects = []
+        updated = []
+        for obj in queryset:
+            if not obj.timeallocation_set.filter(semester=currentsemester):
```
Comment (jnation3406) [inline observation_portal/proposals/admin.py:148]: .exists() on querysets you want to evaluate to truthy value

File: observation_portal/proposals/serializers.py [hunk 4]
```diff
@@ -53,7 +53,7 @@ class ProposalInviteResponseSerializer(serializers.Serializer):
 class MembershipSerializer(serializers.ModelSerializer):
     class Meta:
         model = Membership
-        fields = ('id', 'proposal', 'role', 'user', 'time_limit')
+        fields = ('id', 'proposal', 'role', 'user', 'time_limit','active')
```
Comment (jnation3406) [inline observation_portal/proposals/serializers.py:56]: active is a field on the proposal, not on the membership... Why are you adding it here?

File: observation_portal/proposals/templates/admin/rollover_selected.html [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+{% if updated %}
+These proposals will not be updated because they already have time allocations for next semester:
+{% for proposal in updated %}
+<ul>
+  <li>{{ proposal.id }} - {{proposal.title}}</li>
+</ul>
+{% endfor %}
+{% endif %}
+
+{% endblock %}
```
Comment (jnation3406) [inline observation_portal/proposals/templates/admin/rollover_selected.html:46]: newline at the eof

File: observation_portal/proposals/tests.py [hunk 6]
```diff
@@ -396,3 +396,15 @@ def test_clean_time_allocation_instrument_type_overlap_fails(self):
...
+        self.client.post(
+            reverse('admin:proposals_proposal_changelist'),
+            data={'action': 'rollover_selected', 'apply':'Apply', '_selected_action': [str(proposal.pk) for proposal in self.proposals]},
+            follow=True
+        )
+        for proposal in self.proposals:
+            if proposal == self.proposal:
+                self.assertEqual(TimeAllocation.objects.filter(proposal=proposal, semester=self.future_semester).count(), 1)
+            else:
+                self.assertEqual(TimeAllocation.objects.filter(proposal=proposal, semester=self.future_semester).count(), 0)
```
Comment (jnation3406) [inline observation_portal/proposals/tests.py:410]: newline eof

File: observation_portal/settings.py [hunk 7]
```diff
@@ -199,8 +199,13 @@ def get_dramqtiq_broker_url():
 CORS_ORIGIN_ALLOW_ALL = True
 CORS_ALLOW_CREDENTIALS = True
 CORS_URLS_REGEX = r'^/(api|accounts|media)/.*$|^/o/.*'
-CORS_ORIGIN_WHITELIST = get_list_from_env('CORS_ORIGIN_WHITELIST')
-CSRF_TRUSTED_ORIGINS = get_list_from_env('CSRF_TRUSTED_ORIGINS')
+CORS_ALLOW_CREDENTIALS = True
```
Comment (jnation3406) [inline observation_portal/settings.py:202]: This line already exists a few lines up

File: observation_portal/settings.py [hunk 8]
```diff
@@ -199,8 +199,13 @@ def get_dramqtiq_broker_url():
...
+CORS_ORIGIN_WHITELIST = [
```
Comment (jnation3406) [inline observation_portal/settings.py:203]: I don't think we want this defined the observation portal library. These values are deployment dependent, so they will vary between users.

File: observation_portal/settings.py [hunk 9]
```diff
@@ -442,7 +447,7 @@ def get_dramqtiq_broker_url():
 TEST_RUNNER = 'observation_portal.test_runner.MyDiscoverRunner'
 
 try:
-    from local_settings import *  # noqa
+    from .local_settings import *  # noqa
```
Comment (jnation3406) [inline observation_portal/settings.py:450]: Does this change anything - I would think both reference a local settings file in your root directory...

**Review by jnation3406 — CHANGES_REQUESTED** [review]
Looks good, a few changes for clarity.

File: observation_portal/settings.py (hunk 7 above)
Comment (zemogle) [reply observation_portal/settings.py:202]: Sorry about all these changes to `settings.py` That was unintentional. I've reverted them.

File: observation_portal/proposals/serializers.py (hunk 4 above)
Comment (zemogle) [reply observation_portal/proposals/serializers.py:56]: I have no idea where this comes from. I have no reason to add this.

File: observation_portal/proposals/admin.py (hunk 1 above)
Comment (zemogle) [reply observation_portal/proposals/admin.py:130]: Done

File: observation_portal/proposals/admin.py (hunk 2 above)
Comment (zemogle) [reply observation_portal/proposals/admin.py]: Good call. I've implemented exactly that.

File: observation_portal/settings.py (hunk 9 above)
Comment (zemogle) [reply observation_portal/settings.py]: The original does not work for me but nevertheless, I had not intended to commit that. Reverted.

**Comment by zemogle** [comment]
Thanks for the review comments @jnation3406 I've fixed up the code, if you'd like to take another look 👍

**Review by jnation3406 — APPROVED** [review]
Looks good, just address the codacy issues (trailing whitespace on ln 143 and 153 in proposals/admin.py) and we can merge it.

**Comment by zemogle** [comment]
Thanks for checking that. I didn't notice that. The whitespace is now gone.

**Pull request merged**